allauth/socialaccount/providers/metamask/views.py

```diff
@@ -0,0 +1,71 @@
+import json
+from django.http import JsonResponse
+import random
+import string
+from django.utils.translation import ugettext_lazy as _
+from django.views.decorators.csrf import csrf_exempt
+
+from django.core.exceptions import ImproperlyConfigured
+from django.contrib.auth import logout
+
+from allauth.socialaccount import app_settings, providers
+from allauth.socialaccount.helpers import (
+    complete_social_login,
+    render_authentication_error,
+)
+from allauth.socialaccount.models import SocialLogin, SocialToken
+from .provider import MetamaskProvider
+from django.views.decorators.http import require_http_methods
+
+# web3 declarations
+from web3 import Web3
+from eth_account.messages import encode_defunct
+from django.utils.translation import ugettext_lazy as _
+
+@csrf_exempt
+@require_http_methods(["GET","POST"])
+def login_api(request):
+    extra_data = request.body.decode('utf-8')
+    data = json.loads(extra_data)
+    request.uid = data["account"]
+    request.process = data["process"]
+    if "login_token" in data.keys():
+        request.session['login_token'] = data["login_token"]
+    settings = app_settings.PROVIDERS.get(MetamaskProvider.id, {})
+    provider = providers.registry.by_id(MetamaskProvider.id, request)
+    url = settings.get("URL", "https://cloudflare-eth.com/")
+    port = settings.get("PORT", 80 )
+    w3 = Web3(Web3.HTTPProvider(url+':'+str(port)))
+    if request.process == 'token':
+        token = ''.join(random.SystemRandom().choice(string.ascii_uppercase + string.digits) for i in range(32))
+        request.session['login_token'] = token
+        app = provider.get_app(request)
+        expires_at = None
+        SocialToken.objects.all().filter(account__user__username=data["account"]).delete()
+        storetoken = SocialToken(
+            app=app, token=token, expires_at=expires_at
+        )
+        login = providers.registry.by_id(MetamaskProvider.id, request).sociallogin_from_response(request, data)
+        login.state = SocialLogin.state_from_request(request)
+        login.token = storetoken
+        complete_social_login(request, login)
+        logout(request)
```

Comment:
Why logout?

Reply:
This is probably an ugly hack, but basically, I need the account created but the user to not be logged in at this point in the login flow. This is so I can associate a SocialToken object with the account for the verification step so I can store the original signature string. I based this all off the Persona provider and I don't exactly understand immediately how to create an account without logging in - so I copied the code to create an account and log them in, and immediately log them out before sending them back to the ajax functions achieved the desired goal. 

Please let me know if there's a better way to do this.

Reply:
Please let me know about this.